### ISSUE TITLE
Clean up and robustify USB communication

### ISSUE THREAD

**Issue by madcowswe** [odriverobotics/ODrive#17]
It should be able to be configured and commissioned over USB, and it should be able to take runtime commands from USB.

If that means that we have a G-Code parser on board, or if we hook it up through some python on the PC that then speaks some binary protocol is yet to be determined. Feel free to add suggestions.

**Comment by dirk-makerhafen** [comment]
I added a simple get/set usb interface + a python code generator to create an api + a javascript webinterface with live plotting. 

I also added some code for monitoring/live plotting: my idea was this: because there is alway a compromise between "how many values" and "how often" , i added  monitoring_slots, 20 for now. The usb-api exposes a function like "monitor value xy on slot 1". If you then call "o(utput) 2" you get the first 2 monitored values. So a client can configure it to have important values on the first few slots, and read these with kHz, and less important values later in the array, so maybe every once a second it can call "o 20" to get all monitored values. The ODrive.py code makes use of these functions and automatically assigns and reads monitoring slots if you choose to monitor some value with some frequency. 

Pull request: https://github.com/madcowswe/ODriveFirmware/pull/19

**Comment by alexisdal** [comment]
iwth master commit b0e57908ce5612043f9de9885dc72b257ac80cb9
i managed to get USB interface unresponsive with motor not moving anymore by using the following script

```
#!/bin/bash
while(true); do
   num_points="9999"
   echo -en "p 0 $num_points 0 0\n" > /dev/ttyACM0
   sleep 3.0
   echo -en "p 0 -$num_points 0 0\n" > /dev/ttyACM0
   sleep 3.0
done
```

(basically going back and forth)

i get the usb interface to stop responding after about 1 or 2 minutes

**Comment by alexisdal** [comment]
i think it is not relevant but i can post specific motor params in low_level.c along with engine and encoder specs if that help. plz advise

**Comment by madcowswe** [comment]
Okay, so for right now, here's the plan:

- We wanna keep the ASCII style comms for now.
- Main thing to improve is presenting an API that has symbols for the parameters instead of numbers
- We make all parameters use getters/setters (this is important for setting computed parameters, like control bandwith)
- We store a table of structs containing either:
  - {"name_string", TYPE_AS_ENUM, address_of_getter (or NULL), address_of_setter (or NULL)}
  - {"name_string", CONTAINER_SPECIAL_ENUM, subtree_ptr, subtree_len}
  - The last two elements would be a union.

**Comment by alexisdal** [comment]
i just reproduced the issue using 

* odrive board v3.2
* turniny aerodrive 5065 236
* a simple lasercut hardward jig to hold the motor and the encoder.
* a rohde&shwarz lab power supply (settings on 24v / max 4a)
* a R47 resistor connected on AUX (since i am using a power supply instead of a battery, my understanding was that i should use a "big" resistor connected on AUX)
* lastest master code (06047cb79139837de582afbf32202c237e1b5627) updated by Oskar 6 days ago. no modification at all.
* the same code snippet as above (basically sending "p 0 9999 0 0\n" and "p 0 -9999 0 0\n" back and forth every 3 seconds)

After about 20~30 seconds, the motor stops moving and usb interface no longer responds. the only way to work with the board is tostop sending request, turn off and on power supply, wait for the init sequence to be over. 

I don't care much about the API interface as long as it is simple and stable. what madcowswe suggests above is very fine for me. I have not tried the step/dir hardware interface yet.

Basically the symptom is the USB interface is not usable.

Should I use another brand? 
Should I use the step/dir interface instead of USB ?
Does anyone reproduce the issue?

**Comment by madcowswe** [comment]
Improved struct contents:
{"name_string", TYPE_AS_ENUM, address_of_variable (or general void*), NULL (or general callback)}

**Comment by madcowswe** [comment]
#52 addresses this.
